### PR TITLE
[Keyboard] Add/Update Pimoroni keybow 2040

### PR DESCRIPTION
**Pull request by willoucom**
This is a PR to try to continue the work done on https://github.com/qmk/qmk_firmware/pull/21481

I am not the original author of the firmware.

## Description

This is a continuation, I tries to update files according to the last unresolved request changes.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/pimoroni/keybow2040/readme.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    but this QMK firmware has different USB IDs to avoid confusing drivers:
+    VID = 0x5069 ('Pi'), PID = 0x4784 (4 'x' 4)
+
+
+## resources
+
+Here are some links to the information I used when configuring QMK for the Keybow 2040.
+
+  * [Piomoroni Keybow 2040 product page](https://shop.pimoroni.com/products/keybow-2040)
+  * [Keybow 2040 assembly instructions](https://learn.pimoroni.com/article/assembling-keybow-2040)
```
Comment (zvecr) [inline keyboards/pimoroni/keybow2040/readme.md:64]:
```suggestion
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the top left key and plug in the keyboard
* **Physical reset button**: Hold down BOOTSEL while pressing RESET
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/pimoroni/keybow2040/info.json
[the remark this answers is not in the document]
Comment (willoucom) [reply keyboards/pimoroni/keybow2040/info.json:1]: done

File: keyboards/pimoroni/keybow2040/rules.mk
[the remark this answers is not in the document]
Comment (willoucom) [reply keyboards/pimoroni/keybow2040/rules.mk:1]: done

**Comment by willoucom** [comment]
Hello @zvecr 
Thanks for your review, I fixed all the issues.

File: keyboards/pimoroni/keybow2040/config.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+
+#define IS31FL3731_I2C_ADDRESS_1 0b1110100
```
Comment (zvecr) [inline keyboards/pimoroni/keybow2040/config.h:7]:
```suggestion
#define IS31FL3731_I2C_ADDRESS_1 IS31FL3731_I2C_ADDRESS_GND
```

File: keyboards/pimoroni/keybow2040/keyboard.json [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            "hue_wave": true,
+            "jellybean_raindrops": true,
+            "pixel_flow": true,
+            "pixel_fractal": true,
+            "pixel_rain": true,
+            "rainbow_beacon": true,
+            "rainbow_moving_chevron": true,
+            "rainbow_pinwheels": true,
+            "raindrops": true,
+            "solid_color": true,
```
Comment (zvecr) [inline keyboards/pimoroni/keybow2040/keyboard.json:53]:
```suggestion
```

File: keyboards/pimoroni/keybow2040/keybow2040.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "quantum.h"
+
+#ifdef RGB_MATRIX_ENABLE
+/*
+ * the LED addresses must match the wiring in the Keybow 2040 schematic
+ */
```
Comment (zvecr) [inline keyboards/pimoroni/keybow2040/keybow2040.c:8]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Missed a few things on the last review.

**Comment by willoucom** [comment]
> Missed a few things on the last review.

No problem

File: keyboards/pimoroni/keybow2040/config.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+#pragma once
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+
+#define IS31FL3731_I2C_ADDRESS_1 IS31FL3731_I2C_ADDRESS_GND
+
+#define I2C_DRIVER I2CD0
+#define I2C1_SDA_PIN GP4
+#define I2C1_SCL_PIN GP5
+#define I2C1_CLOCK_SPEED 100000
```
Comment (drashna) [inline keyboards/pimoroni/keybow2040/config.h:12]: Only 100kHz?   I know the 3731 can handle 400kHz.

**Review by drashna — APPROVED** [review]

**Comment by willoucom** [comment]
Hello @zvecr 
Is everything ok ?

**Comment by willoucom** [comment]
Hi @lesshonor 
Thanks for your review

**Review by tris203 — APPROVED** [review]

**Review by jamesfoley — CHANGES_REQUESTED** [review]
Works great on my Keybow2040.

Could I request that the enabled animations gets changed to the following:
```
{
  "alphas_mods": true,
  "band_pinwheel_sat": true,
  "band_pinwheel_val": true,
  "band_sat": true,
  "band_spiral_sat": true,
  "band_spiral_val": true,
  "band_val": true,
  "breathing": true,
  "cycle_all": true,
  "cycle_left_right": true,
  "cycle_out_in": true,
  "cycle_out_in_dual": true,
  "cycle_pinwheel": true,
  "cycle_spiral": true,
  "cycle_up_down": true,
  "digital_rain": true,
  "dual_beacon": true,
  "gradient_left_right": true,
  "gradient_up_down": true,
  "hue_breathing": true,
  "hue_pendulum": true,
  "hue_wave": true,
  "jellybean_raindrops": true,
  "multisplash": true,
  "pixel_flow": true,
  "pixel_fractal": true,
  "pixel_rain": true,
  "rainbow_beacon": true,
  "rainbow_moving_chevron": true,
  "rainbow_pinwheels": true,
  "raindrops": true,
  "solid_multisplash": true,
  "solid_reactive": true,
  "solid_reactive_cross": true,
  "solid_reactive_multicross": true,
  "solid_reactive_multinexus": true,
  "solid_reactive_multiwide": true,
  "solid_reactive_nexus": true,
  "solid_reactive_simple": true,
  "solid_reactive_wide": true,
  "solid_splash": true,
  "splash": true,
  "typing_heatmap": true
}
```

Otherwise the selected animation in VIA doesn't match the one running on the Keybow2040. It also fixes a couple of changed animation names.

The below definition can be used to get it working in VIA:

```
{
    "name": "Pimoroni Keybow 2040",
    "vendorId": "0x5069",
    "productId": "0x4784",
    "matrix": {
        "rows": 5,
        "cols": 4
    },
    "menus": [ "qmk_rgb_matrix" ],
    "keycodes": [ "qmk_lighting" ],
    "layouts": {
        "keymap": [
            [
                {
                    "x": 2.5,
                    "w": 0.5,
                    "h": 0.5,
                    "w2": 1
                },
                "4,0"
            ],
            [
                {
                    "y": -0.5
                },
                "0,0",
                "0,1",
                "0,2",
                "0,3"
            ],
            [
                "1,0",
                "1,1",
                "1,2",
                "1,3"
            ],
            [
                "2,0",
                "2,1",
                "2,2",
                "2,3"
            ],
            [
                "3,0",
                "3,1",
                "3,2",
                "3,3"
            ]
        ]
    }
}
```

File: keyboards/pimoroni/keybow2040/config.h
Comment (jamesfoley) [inline keyboards/pimoroni/keybow2040/config.h:1]: Could we add the following to get RGB turned off during USB suspend?

```
#ifdef RGB_MATRIX_ENABLE
    #define RGB_MATRIX_SLEEP
#endif
```

Working on Windows but doesn't on my Mac for some reason, though not sure thats an issue specifically related to the Keybow2040.

Comment (fauxpark) [reply keyboards/pimoroni/keybow2040/config.h:1]: macOS will suspend USB devices after a short time, it's not instant like on Windows.

Comment (jamesfoley) [reply keyboards/pimoroni/keybow2040/config.h:1]: Ah cool okay, thats good to know.

Comment (willoucom) [reply keyboards/pimoroni/keybow2040/config.h:1]: I tried to add those lines into config.h and I don't see a difference, how can I trigger an usb suspend on windows ?

Comment (jamesfoley) [reply keyboards/pimoroni/keybow2040/config.h:1]: With the Keybow 2040 plugged into a motherboard configured to supply power to USB ports after shutdown the RGB will stay active. With `RGB_MATRIX_SLEEP` defined the RGB sleeps at shutdown regardless of USB power.

Comment (willoucom) [reply keyboards/pimoroni/keybow2040/config.h:1]: Thanks, I got it to work on windows, can you test it on MacOS ?

Comment (jamesfoley) [reply keyboards/pimoroni/keybow2040/config.h:1]: Can confirm that the RGB turns off instantly on Windows, and after a short delay on macOS.

As a side note, it seems that the issue I had on macOS was that the Keybow was plugged into a Caldigit TS3+ dock, and a quick Google found:

> Three of the USB ports should turn off entirely when the computer is disconnected or not active.
> These ports are the two USB-C ports, and the rear left-most USB-A port.

Swapped the Keybow to one of the above mentioned ports and now when the Mac is put to sleep, after a few minutes the RGB turns off on the Keybow. I have no idea why its USB port specific, but it is what it is.

**Comment by sandyjmacdonald** [comment]
Hahaha. I wish I'd thought to check open PRs before launching into #24276. Really nice work on this, @willoucom and @fanf2! Thanks to @keyboard-magpie for alerting me.

That being said, it all looks very similar to what I wrote. If you would like someone on the Pimoroni team (albeit not full-time now) to contribute to this, @willoucom, then I'd be more than happy to help. Or even take it over, if you don't want to be lumbered with it... totally up to you. 🙂

I notice that the hyperlink to the Keybow 2040 product page is borked, and you'd be more than welcome to use my photo that's on the product page if you like: [https://i.imgur.com/WDGDOkb.jpeg](https://i.imgur.com/WDGDOkb.jpeg)

The [default keymap I added here](https://github.com/sandyjmacdonald/qmk_firmware/blob/develop/keyboards/keybow/keybow_2040/keymaps/default/keymap.c) is a little more advanced than the one here, with a second layer to control the LED animations. Again, feel free to use it if it's helpful.

**Comment by Gadgetoid** [comment]
> I do like the use of the boot button as an additional mod key!

Sad v1 keybow 2040 noises

**Comment by willoucom** [comment]
Hello @sandyjmacdonald 
I like your suggestions and have modified the behavior of the layers to reflect your changes.
I've also added you as a contributor to this PR.

Can you test these modifications and tell me if everything is correct for you?

Thank you very much for your help.

**Review by keyboard-magpie — APPROVED** [review]

**Pull request merged**